File: langs/go/validator.go
package dotlyte

import (
	"fmt"
	"net"
	"net/url"
	"regexp"
	"strconv"
	"strings"
)

// SchemaRule defines validation constraints for a single config key.
type SchemaRule struct {
	Type      string         // "string", "number", "boolean", "array", "object"
	Required  bool           // key must exist
	Format    string         // built-in format: url, email, ip, ipv4, ipv6, hostname, port, uuid, date
	Pattern   string         // regex pattern for string values
	Enum      []any          // allowed values
	Min       *float64       // minimum (for numbers)
	Max       *float64       // maximum (for numbers)
	Default   any            // default value if missing
	Sensitive bool           // marks key as containing secrets
	Doc       string         // documentation string
	Validator func(any) bool // custom validation function
}

// DotlyteSchema maps config keys to their SchemaRules.
type DotlyteSchema map[string]*SchemaRule

// SchemaViolation describes a single schema validation failure.
type SchemaViolation struct {
	Key     string
	Message string
	Rule    string // which rule failed: "required", "type", "format", "enum", "min", "max", "pattern", "validator"
}

func (v *SchemaViolation) String() string {
	return fmt.Sprintf("%s: %s (%s)", v.Key, v.Message, v.Rule)
}

// ValidateSchema checks config data against a schema, returning all violations.
func ValidateSchema(data map[string]any, schema DotlyteSchema, strict bool) []SchemaViolation {
	var violations []SchemaViolation

	for key, rule := range schema {
		val := getNestedValue(data, key)

		if val == nil {
			if rule.Required {
				violations = append(violations, SchemaViolation{
					Key:     key,
					Message: fmt.Sprintf("required key '%s' is missing", key),
					Rule:    "required",
				})
			}
			continue
		}

		// Type check
		if rule.Type != "" && !checkType(val, rule.Type) {
			violations = append(violations, SchemaViolation{
				Key:     key,
				Message: fmt.Sprintf("expected type '%s', got %T", rule.Type, val),
				Rule:    "type",
			})
		}

		// Format check
		if rule.Format != "" {
			if s, ok := val.(string); ok {
				if !checkFormat(s, rule.Format) {
					violations = append(violations, SchemaViolation{
						Key:     key,
						Message: fmt.Sprintf("value '%s' does not match format '%s'", s, rule.Format),
						Rule:    "format",
					})
				}
			}
		}

		// Pattern check
		if rule.Pattern != "" {
			if s, ok := val.(string); ok {
				if matched, _ := regexp.MatchString(rule.Pattern, s); !matched {
					violations = append(violations, SchemaViolation{
						Key:     key,
						Message: fmt.Sprintf("value '%s' does not match pattern '%s'", s, rule.Pattern),
						Rule:    "pattern",
					})
				}
			}
		}

		// Enum check
		if len(rule.Enum) > 0 {
			found := false
			for _, allowed := range rule.Enum {
				if fmt.Sprintf("%v", val) == fmt.Sprintf("%v", allowed) {
					found = true
					break
				}
			}
			if !found {
				violations = append(violations, SchemaViolation{
					Key:     key,
					Message: fmt.Sprintf("value '%v' is not one of allowed values: %v", val, rule.Enum),
					Rule:    "enum",
				})
			}
		}

		// Min/Max checks
		if rule.Min != nil || rule.Max != nil {
			num := toFloat64(val)
			if num != nil {
				if rule.Min != nil && *num < *rule.Min {
					violations = append(violations, SchemaViolation{
						Key:     key,
						Message: fmt.Sprintf("value %v is less than minimum %v", val, *rule.Min),
						Rule:    "min",
					})
				}
				if rule.Max != nil && *num > *rule.Max {
					violations = append(violations, SchemaViolation{
						Key:     key,
						Message: fmt.Sprintf("value %v is greater than maximum %v", val, *rule.Max),
						Rule:    "max",
					})
				}
			}
		}

		// Custom validator
		if rule.Validator != nil && !rule.Validator(val) {
			violations = append(violations, SchemaViolation{
				Key:     key,
				Message: fmt.Sprintf("custom validation failed for key '%s'", key),
				Rule:    "validator",
			})
		}
	}

	// Strict mode: reject keys not in schema
	if strict {
		flatKeys := flattenKeys(data, "")
		for _, k := range flatKeys {
			if _, exists := schema[k]; !exists {
				violations = append(violations, SchemaViolation{
					Key:     k,
					Message: fmt.Sprintf("unknown key '%s' (strict mode)", k),
					Rule:    "strict",
				})
			}
		}
	}

	return violations
}

// ApplySchemaDefaults fills in missing keys with schema default values.
func ApplySchemaDefaults(data map[string]any, schema DotlyteSchema) map[string]any {
	result := make(map[string]any)
	for k, v := range data {
		result[k] = v
	}

	for key, rule := range schema {
		if rule.Default != nil && getNestedValue(result, key) == nil {
			setNestedValue(result, key, rule.Default)
		}
	}

	return result
}

// GetSensitiveKeys returns all keys marked as sensitive in the schema.
func GetSensitiveKeys(schema DotlyteSchema) []string {
	var keys []string
	for key, rule := range schema {
		if rule.Sensitive {
			keys = append(keys, key)
		}
	}
	return keys
}

func getNestedValue(data map[string]any, key string) any {
	parts := strings.Split(key, ".")
	var current any = data
	for _, part := range parts {
		m, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		current, ok = m[part]
		if !ok {
			return nil
		}
	}
	return current
}

func setNestedValue(data map[string]any, key string, value any) {
	parts := strings.Split(key, ".")
	current := data
	for _, part := range parts[:len(parts)-1] {
		if _, ok := current[part]; !ok {
			current[part] = make(map[string]any)
		}
		if m, ok := current[part].(map[string]any); ok {
			current = m
		} else {
			m := make(map[string]any)
			current[part] = m
			current = m
		}
	}
	current[parts[len(parts)-1]] = value
}

func flattenKeys(data map[string]any, prefix string) []string {
	var keys []string
	for k, v := range data {
		fullKey := k
		if prefix != "" {
			fullKey = prefix + "." + k
		}
		if m, ok := v.(map[string]any); ok {
			keys = append(keys, flattenKeys(m, fullKey)...)
		} else {
			keys = append(keys, fullKey)
		}
	}
	return keys
}

func checkType(val any, expected string) bool {
	switch expected {
	case "string":
		_, ok := val.(string)
		return ok
	case "number":
		return toFloat64(val) != nil
	case "boolean":
		_, ok := val.(bool)
		return ok
	case "array":
		_, ok := val.([]any)
		return ok
	case "object":
		_, ok := val.(map[string]any)
		return ok
	}
	return true
}

var (
	emailRegex = regexp.MustCompile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)
	uuidRegex  = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$`)
	dateRegex  = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)
)

func checkFormat(val, format string) bool {
	switch format {
	case "url":
		u, err := url.ParseRequestURI(val)
		return err == nil && u.Scheme != "" && u.Host != ""
	case "email":
		return emailRegex.MatchString(val)
	case "ip":
		return net.ParseIP(val) != nil
	case "ipv4":
		ip := net.ParseIP(val)
		return ip != nil && ip.To4() != nil
	case "ipv6":
		ip := net.ParseIP(val)
		return ip != nil && ip.To4() == nil
	case "hostname":
		matched, _ := regexp.MatchString(`^[a-zA-Z0-9]([a-zA-Z0-9\-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9\-]*[a-zA-Z0-9])?)*$`, val)
		return matched
	case "port":
		p, err := strconv.Atoi(val)
		return err == nil && p >= 1 && p <= 65535
	case "uuid":
		return uuidRegex.MatchString(val)
	case "date":
		return dateRegex.MatchString(val)
	default:
		// Treat format as regex pattern
		matched, err := regexp.MatchString(format, val)
		return err == nil && matched
	}
}

func toFloat64(val any) *float64 {
	var f float64
	switch v := val.(type) {
	case int:
		f = float64(v)
	case int64:
		f = float64(v)
	case float64:
		f = v
	case float32:
		f = float64(v)
	default:
		return nil
	}
	return &f
}

File: langs/rust/src/validator.rs
//! Schema validation for DOTLYTE v2.

use std::collections::HashMap;

use regex::Regex;
use serde_json::Value;

use crate::errors::{SchemaViolation, DotlyteError};

/// A single schema rule for one config key.
#[derive(Debug, Clone, Default)]
pub struct SchemaRule {
    /// Expected type: "string", "number", "boolean", "array", "object"
    pub value_type: Option<String>,
    /// Whether the key must exist.
    pub required: bool,
    /// Built-in format: "url", "email", "ip", "ipv4", "ipv6", "port", "uuid", "date"
    pub format: Option<String>,
    /// Regex pattern for string values.
    pub pattern: Option<String>,
    /// Allowed values.
    pub enum_values: Option<Vec<Value>>,
    /// Minimum (for numbers).
    pub min: Option<f64>,
    /// Maximum (for numbers).
    pub max: Option<f64>,
    /// Default value if missing.
    pub default: Option<Value>,
    /// Marks key as containing secrets.
    pub sensitive: bool,
    /// Documentation string.
    pub doc: Option<String>,
}

/// Schema mapping keys to rules.
pub type DotlyteSchema = HashMap<String, SchemaRule>;

/// Validate config data against a schema.
pub fn validate_schema(
    data: &serde_json::Map<String, Value>,
    schema: &DotlyteSchema,
    strict: bool,
) -> Vec<SchemaViolation> {
    let mut violations = Vec::new();

    for (key, rule) in schema {
        let val = get_nested(data, key);

        if val.is_none() || val == Some(&Value::Null) {
            if rule.required {
                violations.push(SchemaViolation {
                    key: key.clone(),
                    message: format!("required key '{key}' is missing"),
                    rule: "required".to_string(),
                });
            }
            continue;
        }
        let val = val.unwrap();

        // Type check
        if let Some(ref expected) = rule.value_type {
            if !check_type(val, expected) {
                violations.push(SchemaViolation {
                    key: key.clone(),
                    message: format!("expected type '{expected}', got {}", value_type_name(val)),
                    rule: "type".to_string(),
                });
            }
        }

        // Format check
        if let Some(ref format) = rule.format {
            if let Some(s) = val.as_str() {
                if !check_format(s, format) {
                    violations.push(SchemaViolation {
                        key: key.clone(),
                        message: format!("value '{s}' does not match format '{format}'"),
                        rule: "format".to_string(),
                    });
                }
            }
        }

        // Pattern check
        if let Some(ref pattern) = rule.pattern {
            if let Some(s) = val.as_str() {
                if let Ok(re) = Regex::new(pattern) {
                    if !re.is_match(s) {
                        violations.push(SchemaViolation {
                            key: key.clone(),
                            message: format!("value '{s}' does not match pattern '{pattern}'"),
                            rule: "pattern".to_string(),
                        });
                    }
                }
            }
        }

        // Enum check
        if let Some(ref allowed) = rule.enum_values {
            if !allowed.contains(val) {
                violations.push(SchemaViolation {
                    key: key.clone(),
                    message: format!("value {val} not in allowed values: {allowed:?}"),
                    rule: "enum".to_string(),
                });
            }
        }

        // Min/Max checks
        if let Some(num) = val.as_f64() {
            if let Some(min) = rule.min {
                if num < min {
                    violations.push(SchemaViolation {
                        key: key.clone(),
                        message: format!("value {num} is less than minimum {min}"),
                        rule: "min".to_string(),
                    });
                }
            }
            if let Some(max) = rule.max {
                if num > max {
                    violations.push(SchemaViolation {
                        key: key.clone(),
                        message: format!("value {num} is greater than maximum {max}"),
                        rule: "max".to_string(),
                    });
                }
            }
        }
    }

    // Strict mode
    if strict {
        let flat = flatten_keys_json(data, "");
        for k in flat {
            if !schema.contains_key(&k) {
                violations.push(SchemaViolation {
                    key: k.clone(),
                    message: format!("unknown key '{k}' (strict mode)"),
                    rule: "strict".to_string(),
                });
            }
        }
    }

    violations
}

/// Apply schema defaults: fill in missing keys with defaults.
pub fn apply_schema_defaults(
    data: &mut serde_json::Map<String, Value>,
    schema: &DotlyteSchema,
) {
    for (key, rule) in schema {
        if let Some(ref default) = rule.default {
            if get_nested(data, key).is_none() {
                set_nested(data, key, default.clone());
            }
        }
    }
}

/// Get sensitive keys from schema.
pub fn get_sensitive_keys(schema: &DotlyteSchema) -> Vec<String> {
    schema
        .iter()
        .filter(|(_, rule)| rule.sensitive)
        .map(|(key, _)| key.clone())
        .collect()
}

/// Assert valid — returns error if violations exist.
pub fn assert_valid(
    data: &serde_json::Map<String, Value>,
    schema: &DotlyteSchema,
    strict: bool,
) -> crate::Result<()> {
    let violations = validate_schema(data, schema, strict);
    if violations.is_empty() {
        Ok(())
    } else {
        Err(DotlyteError::ValidationError { violations })
    }
}

fn get_nested<'a>(data: &'a serde_json::Map<String, Value>, key: &str) -> Option<&'a Value> {
    let parts: Vec<&str> = key.split('.').collect();
    if parts.is_empty() {
        return None;
    }
    let mut current: &Value = data.get(parts[0])?;
    for part in &parts[1..] {
        match current {
            Value::Object(map) => {
                current = map.get(*part)?;
            }
            _ => return None,
        }
    }
    Some(current)
}

fn set_nested(data: &mut serde_json::Map<String, Value>, key: &str, value: Value) {
    let parts: Vec<&str> = key.split('.').collect();
    if parts.len() == 1 {
        data.insert(parts[0].to_string(), value);
        return;
    }

    // Navigate/create intermediate objects as a Value tree
    let mut root = Value::Object(data.clone());
    {
        let mut current = &mut root;
        for part in &parts[..parts.len() - 1] {
            if !current.is_object() {
                *current = Value::Object(serde_json::Map::new());
            }
            let map = current.as_object_mut().unwrap();
            if !map.contains_key(*part) {
                map.insert(part.to_string(), Value::Object(serde_json::Map::new()));
            }
            current = map.get_mut(*part).unwrap();
        }
        if let Some(map) = current.as_object_mut() {
            map.insert(parts.last().unwrap().to_string(), value);
        }
    }
    if let Value::Object(map) = root {
        *data = map;
    }
}

fn flatten_keys_json(data: &serde_json::Map<String, Value>, prefix: &str) -> Vec<String> {
    let mut keys = Vec::new();
    for (k, v) in data {
        let full_key = if prefix.is_empty() {
            k.clone()
        } else {
            format!("{prefix}.{k}")
        };
        if let Value::Object(ref map) = v {
            keys.extend(flatten_keys_json(map, &full_key));
        } else {
            keys.push(full_key);
        }
    }
    keys
}

fn check_type(val: &Value, expected: &str) -> bool {
    match expected {
        "string" => val.is_string(),
        "number" => val.is_number(),
        "boolean" => val.is_boolean(),
        "array" => val.is_array(),
        "object" => val.is_object(),
        _ => true,
    }
}

fn value_type_name(val: &Value) -> &'static str {
    match val {
        Value::Null => "null",
        Value::Bool(_) => "boolean",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

fn check_format(val: &str, format: &str) -> bool {
    match format {
        "email" => {
            let re = Regex::new(r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$").unwrap();
            re.is_match(val)
        }
        "url" => val.starts_with("http://") || val.starts_with("https://"),
        "ip" | "ipv4" => {
            val.split('.')
                .filter_map(|s| s.parse::<u8>().ok())
                .count()
                == 4
        }
        "ipv6" => val.contains(':') && val.len() > 2,
        "port" => val.parse::<u16>().map_or(false, |p| p >= 1),
        "uuid" => {
            let re =
                Regex::new(r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")
                    .unwrap();
            re.is_match(val)
        }
        "date" => {
            let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
            re.is_match(val)
        }
        _ => {
            // Treat as regex
            Regex::new(format).map_or(true, |re| re.is_match(val))
        }
    }
}
